lib/autoupdate/agent/updater.go

```diff
@@ -0,0 +1,338 @@
+	Remove(ctx context.Context, version string) error
+}
+
+// UserConfig contains user overrides for installing Teleport agents.
+type UserConfig struct {
+	// Proxy address, scheme and port optional.
+	// Overrides existing value if specified.
+	Proxy string
+	// Group identifier for updates (e.g., staging)
+	// Overrides existing value if specified.
+	Group string
+	// URLTemplate for the Teleport tgz download URL
+	// Overrides existing value if specified.
+	URLTemplate string
+	// ForceVersion to the specified version.
+	ForceVersion string
+}
+
+// Enable enables agent updates and attempts an initial update.
+// If the initial update succeeds, auto-updates are enabled and the configuration is persisted.
+// Otherwise, the auto-updates configuration is not changed.
+// This function is idempotent.
+func (u *Updater) Enable(ctx context.Context, userCfg UserConfig) error {
+	// Read configuration from update.yaml and override any new values passed as flags.
+	cfg, err := u.readConfig(u.ConfigPath)
+	if err != nil {
+		return trace.Errorf("failed to read %s: %w", updateConfigName, err)
+	}
+	if userCfg.Proxy != "" {
+		cfg.Spec.Proxy = userCfg.Proxy
+	}
+	if userCfg.Group != "" {
+		cfg.Spec.Group = userCfg.Group
+	}
+	if userCfg.URLTemplate != "" {
+		cfg.Spec.URLTemplate = userCfg.URLTemplate
+	}
+	cfg.Spec.Enabled = true
+	if err := validateUpdatesSpec(&cfg.Spec); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Lookup target version from the proxy.
+	addr, err := libutils.ParseAddr(cfg.Spec.Proxy)
+	if err != nil {
+		return trace.Errorf("failed to parse proxy server address: %w", err)
+	}
+
+	desiredVersion := userCfg.ForceVersion
+	if desiredVersion == "" {
+		resp, err := webclient.Find(&webclient.Config{
+			Context:   ctx,
+			ProxyAddr: addr.Addr,
+			Insecure:  u.InsecureSkipVerify,
+			Timeout:   30 * time.Second,
+			//Group:     cfg.Spec.Group, // TODO(sclevine): add web API
+			Pool: u.Pool,
+		})
+		if err != nil {
+			return trace.Errorf("failed to request version from proxy: %w", err)
+		}
+		desiredVersion, _ = "16.3.0", resp // TODO(sclevine): add web API
+		//desiredVersion := resp.AutoUpdate.AgentVersion
+	}
+
+	if desiredVersion == "" {
+		return trace.Errorf("agent version not available from Teleport cluster")
+	}
+	// If the active version and target don't match, kick off upgrade.
+	if cfg.Status.ActiveVersion != desiredVersion {
+		template := cfg.Spec.URLTemplate
+		if template == "" {
+			template = cdnURITemplate
+		}
+		// Create /var/lib/teleport/versions/X.Y.Z if it does not exist.
+		err = u.Installer.Install(ctx, desiredVersion, template)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg.Status.ActiveVersion = desiredVersion
+		u.Log.InfoContext(ctx, "Target version successfully installed.", "version", desiredVersion)
+	} else {
+		u.Log.InfoContext(ctx, "Target version already installed.", "version", desiredVersion)
+	}
+
+	// Always write the configuration file if enable succeeds.
```

Comment:
Oops, good catch. This logic was intended for `Update`, since we don't want to query the checksum every 10 min on every agent install. `Install` is already idempotent.

I removed the conditional on `Install` and documented the idempotency.